### ISSUE TITLE
sql: dropping objects has a quadratic time complexity in relation to the number of objects

### ISSUE THREAD

**Issue by RichardJCai** [cockroachdb/cockroach#50783]
Dropping 1 table takes 7 roundtrips, 2 takes 16 and 3 takes 27. (See below for ddl analysis snippet)
I think think there is a n^2 relationship between number of objects dropped using a drop statement and the number of objects being dropped.

This comes from the fact that every dropped object results in a call to `createOrUpdateSchemaChangeJob` which does a loop through the the current evalContext's jobs to check if a job has been queued up or not for the drop, each check performs a round trip because it queries for the job (`SELECT payload, progress FROM system.jobs WHERE id = $1`) and queues a job if one doesn't exist.

This means if for the n'th object being dropped, it will do n - 1 queries (each a roundtrip) to find if a job has been queued or not for it.

```
# DropTable
exec
CREATE TABLE t1()
----

count
DROP TABLE t1
----
7

exec
CREATE TABLE t1();
CREATE TABLE t2();
----

count
DROP TABLE t1,t2
----
16

exec
CREATE TABLE t1();
CREATE TABLE t2();
CREATE TABLE t3();
----

count
DROP TABLE t1,t2,t3
----
27
```

**Comment by RichardJCai** [comment]
@lucy-zhang I think this is related to the slowness of the `GRANT *` on multiple tables as well.

**Comment by awoods187** [comment]
This is an amazing find!

**Comment by ajwerner** [comment]
Time to address this TODO:
https://github.com/cockroachdb/cockroach/blob/f1d6bc33439b20c64f2712b237ac09a3c285c10e/pkg/sql/table.go#L86-L87

Keeping a map from table ID to job ID in `(*connExecutor).extraTxnState` does seem like the move. 

https://github.com/cockroachdb/cockroach/blob/f1d6bc33439b20c64f2712b237ac09a3c285c10e/pkg/sql/conn_executor.go#L924-L929

**Comment by RichardJCai** [comment]
> Time to address this TODO:
> https://github.com/cockroachdb/cockroach/blob/f1d6bc33439b20c64f2712b237ac09a3c285c10e/pkg/sql/table.go#L86-L87
> 
> Keeping a map from table ID to job ID in `(*connExecutor).extraTxnState` does seem like the move.
> 
> https://github.com/cockroachdb/cockroach/blob/f1d6bc33439b20c64f2712b237ac09a3c285c10e/pkg/sql/conn_executor.go#L924-L929

@ajwerner have you began any work on this given that you're assigned to https://github.com/cockroachdb/cockroach/issues/50165 which seems to have the same root issue? If not, mind if I work on it?

**Comment by lucy-zhang** [comment]
@RichardJCai Thanks, go ahead.

**Comment by RichardJCai** [comment]
Fixed by #50848
